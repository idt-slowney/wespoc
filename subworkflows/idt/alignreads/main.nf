include { BWAMEM2_MEM }                                 from '../../../modules/idt/bwamem2/mem/main'
include { SENTIEON_BWAMEM }                             from '../../../modules/idt/sentieon/bwamem/main'
include { PICARD_MARKDUPLICATES }                       from '../../../modules/idt/picard/markduplicates/main'
include { SENTIEON_DEDUP }                              from '../../../modules/idt/sentieon/dedup/main'

workflow ALIGNREADS {
    take:
    reads                   // channel: [mandatory] [ meta, reads ]
    fasta                   // channel: [mandatory] [ meta, fasta ]
    fasta_fai               // channel: [mandatory] [ meta, fasta_fai ]
    index                   // path: [mandatory] index/*
    sort_bam                // bool: [mandatory] default: true
    aligner                 // str: "bwa2" or "sentieon-bwamem", default: "bwa2"
    skip_dedup              // bool: [mandatory] default: false

    main:
    // Initialize output channels
    dupmarked_bams = Channel.empty()
    reports = Channel.empty()
    ch_versions = Channel.empty()

    reads_to_align = reads
    switch (aligner) {
        case "bwa2":
            BWAMEM2_MEM(reads,  index, fasta, sort_bam)
            aligned_reads = BWAMEM2_MEM.out.bam
            ch_versions = ch_versions.mix(BWAMEM2_MEM.out.versions)

            if (!skip_dedup) {
                PICARD_MARKDUPLICATES(BWAMEM2_MEM.out.bam, fasta, fasta_fai)
                dupmarked_bams = dupmarked_bams.mix(PICARD_MARKDUPLICATES.out.bam.join(PICARD_MARKDUPLICATES.out.bai, failOnDuplicate: true, failOnMismatch: true))
                reports = reports.mix(PICARD_MARKDUPLICATES.out.metrics)
                ch_versions = ch_versions.mix(PICARD_MARKDUPLICATES.out.versions)
            }
            break

        case "sentieon-bwamem":
            SENTIEON_BWAMEM(reads, index, fasta, fasta_fai)
            aligned = SENTIEON_BWAMEM.out.bam
            ch_versions = ch_versions.mix(SENTIEON_BWAMEM.out.versions)

            if (!skip_dedup) {
                SENTIEON_DEDUP( SENTIEON_BWAMEM.out.bam_and_bai, fasta, fasta_fai)
                dupmarked_bams = dupmarked_bams.mix(SENTIEON_DEDUP.out.bam.join(SENTIEON_DEDUP.out.bai, failOnDuplicate: true, failOnMismatch: true))
                reports = reports.mix(SENTIEON_DEDUP.out.metrics)
                ch_versions = ch_versions.mix(SENTIEON_DEDUP.out.versions)
            }
            break

        default:
            error "Unknown aligner: ${aligner}"
    }

    emit:
    aligned_bams    = aligned_reads       // channel: [ [meta], bam ]
    dupmarked_bams                        // channel: [ [meta], bam, bai ]
    reports                               // channel: [ [meta], metrics ]

    versions        = ch_versions         // channel: [ ch_versions.yml ]
}
