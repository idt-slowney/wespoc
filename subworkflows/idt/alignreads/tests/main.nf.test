nextflow_workflow {

    name "Test Subworkflow ALIGNREADS"
    script "../main.nf"
    workflow "ALIGNREADS"
    config './nextflow.config'

    tag "subworkflows_nfcore"
    tag "subworkflows"
    tag "subworkflows/alignreads"
    tag "bwamem2/mem"
    tag "sentieon/bwamem"
    tag "picard/sortsam"
    tag "picard/markduplicates"
    tag "sentieon/dedup"
    tag "bwamem2/index"


    test("bwamem2 on paired end illumina fastq") {
        setup {
            run("BWAMEM2_INDEX") {
                script "../../../../modules/idt/bwamem2/index/main.nf"
                process {
                    """
                    input[0] = Channel.value([ 
                        [ id:'genome' ],
                          file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true)
                    ])
                    """
                }
            }
        }
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false, sample_size:10000 ],
                    [ file(params.test_data['homo_sapiens']['illumina']['test_1_fastq_gz'], checkIfExists: true),
                      file(params.test_data['homo_sapiens']['illumina']['test_2_fastq_gz'], checkIfExists: true)                   ]
                ])
                input[1] = Channel.of([ 
                    [ id:'genome' ],
                      file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true)
                ])
                input[2] = Channel.of([ 
                    [ id:'genome_fai' ],
                      file(params.test_data['homo_sapiens']['genome']['genome_fasta_fai'], checkIfExists: true)
                ])
                input[3] = BWAMEM2_INDEX.out.index
                input[4] = true
                input[5] = 'bwa2'
                input[6] = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    // TODO: Run this test when we acquire Sentieon license
    // test("sentieon-bwamem on paired end illumina fastq") {
    //     setup {
    //         run("SENTIEON_BWAINDEX") {
    //             script "../../../../modules/idt/sentieon/bwaindex/main.nf"
    //             process {
    //                 """
    //                 input[0] = Channel.value([ [ id:'genome' ],
    //                     file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true)
    //                     ])
    //                 """
    //             }
    //         }
    //     }
    //     when {
    //         workflow {
    //             """
    //             input[0] = Channel.of([
    //                 [ id:'test', single_end:false, sample_size:10000 ],
    //                 [
    //                     file(params.test_data['homo_sapiens']['illumina']['test_1_fastq_gz'], checkIfExists: true),
    //                     file(params.test_data['homo_sapiens']['illumina']['test_2_fastq_gz'], checkIfExists: true)
    //                 ]
    //             ])
    //             input[1] = Channel.of([ [ id:'genome' ],
    //                     file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true)
    //                     ])
    //             input[2] = Channel.of([ [ id:'genome_fai' ],
    //                     file(params.test_data['homo_sapiens']['genome']['genome_fasta_fai'], checkIfExists: true)
    //                     ])
    //             input[3] = SENTIEON_BWAINDEX.out.index
    //             input[4] = true
    //             input[5] = 'sentieon-bwamem'
    //             input[6] = false
    //             """
    //         }
    //     }

    //     then {
    //         assertAll(
    //             { assert workflow.success},
    //             { assert snapshot(workflow.out).match()}
    //         )
    //     }
    // }
}
