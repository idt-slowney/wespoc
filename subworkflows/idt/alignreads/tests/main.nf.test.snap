{
    "bwamem2 on paired end illumina fastq": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "sample_size": 10000
                        },
                        "test.aligner.test.Aligned.bam:md5,52b3ae00dba9d4e9cd0513793378eee7"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "sample_size": 10000
                        },
                        "test.aligner.test.DupMarked.bam:md5,68a4d053a5b5c3b1b31bd8935d63e3f4",
                        "test.aligner.test.DupMarked.bai:md5,28c1956bd71100f7cd4e539e4ae73bcb"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "sample_size": 10000
                        },
                        "test.aligner.test.DupMarked.MarkDuplicates.metrics.txt:md5,1e4050d8eb370f3f639ae3d57a32b067"
                    ]
                ],
                "3": [
                    "versions.yml:md5,75f8463b1d94fea851ae5842e54c440d",
                    "versions.yml:md5,d553735d7bf50aa911b9ce84beb5b525"
                ],
                "aligned_bams": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "sample_size": 10000
                        },
                        "test.aligner.test.Aligned.bam:md5,52b3ae00dba9d4e9cd0513793378eee7"
                    ]
                ],
                "dupmarked_bams": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "sample_size": 10000
                        },
                        "test.aligner.test.DupMarked.bam:md5,68a4d053a5b5c3b1b31bd8935d63e3f4",
                        "test.aligner.test.DupMarked.bai:md5,28c1956bd71100f7cd4e539e4ae73bcb"
                    ]
                ],
                "reports": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "sample_size": 10000
                        },
                        "test.aligner.test.DupMarked.MarkDuplicates.metrics.txt:md5,1e4050d8eb370f3f639ae3d57a32b067"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,75f8463b1d94fea851ae5842e54c440d",
                    "versions.yml:md5,d553735d7bf50aa911b9ce84beb5b525"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-09T14:59:10.510773336"
    }
}